libcontainer/container_linux.go

```diff
@@ -518,57 +518,116 @@ func (c *Container) shouldSendMountSources() bool {
+	// We need to send sources if there are non-idmap bind-mounts.
 	for _, m := range c.config.Mounts {
-		if m.IsBind() {
+		if m.IsBindOnly() {
 			return true
 		}
 	}
 
 	return false
 }
 
-func (c *Container) newInitProcess(p *Process, cmd *exec.Cmd, messageSockPair, logFilePair filePair) (*initProcess, error) {
-	cmd.Env = append(cmd.Env, "_LIBCONTAINER_INITTYPE="+string(initStandard))
-	nsMaps := make(map[configs.NamespaceType]string)
-	for _, ns := range c.config.Namespaces {
-		if ns.Path != "" {
-			nsMaps[ns.Type] = ns.Path
+// shouldSendIdmapSources says whether the child process must setup idmap mounts with
+// the mount_setattr already done in the host user namespace.
+func (c *Container) shouldSendIdmapSources() bool {
+	// For the time being we require userns to be in use.
+	if !c.config.Namespaces.Contains(configs.NEWUSER) {
+		return false
+	}
+
+	// nsexec.c mount_setattr() requires CAP_SYS_ADMIN in:
+	// * the user namespace the filesystem was mounted in
+	// * the user namespace we're trying to idmap the mount to
+	// * the owning user namespace of the mount namespace you're currently located in.
+	//
+	// See the comment from Christian Brauner:
+	//	https://github.com/opencontainers/runc/pull/3717#discussion_r1103607972
+	//
+	// Let's just rule out rootless, we don't have those permission in the
+	// rootless case.
+	if c.config.RootlessEUID {
+		return false
+	}
+
+	// We need to send sources if there are idmap bind-mounts.
+	for _, m := range c.config.Mounts {
+		if m.IsIDMap() {
+			return true
 		}
 	}
-	_, sharePidns := nsMaps[configs.NEWPID]
-	data, err := c.bootstrapData(c.config.Namespaces.CloneFlags(), nsMaps, initStandard)
+
+	return false
+}
+
+func cmdAddEnvFds(cmd *exec.Cmd, env string, fds []int) error {
+	fdsJSON, err := json.Marshal(fds)
 	if err != nil {
-		return nil, err
+		return fmt.Errorf("Error creating %v: %w", env, err)
 	}
+	cmd.Env = append(cmd.Env, env+"="+string(fdsJSON))
+	return nil
+}
 
-	if c.shouldSendMountSources() {
-		// Elements on this slice will be paired with mounts (see StartInitialization() and
-		// prepareRootfs()). This slice MUST have the same size as c.config.Mounts.
-		mountFds := make([]int, len(c.config.Mounts))
-		for i, m := range c.config.Mounts {
-			if !m.IsBind() {
-				// Non bind-mounts do not use an fd.
-				mountFds[i] = -1
-				continue
-			}
+func (c *Container) sendSources(cmd *exec.Cmd, messageSockPair filePair) error {
+	if !c.shouldSendMountSources() && !c.shouldSendIdmapSources() {
+		return nil
+	}
 
+	// Elements on these slices will be paired with mounts (see StartInitialization() and
+	// prepareRootfs()). These slices MUST have the same size as c.config.Mounts.
+	mountFds := make([]int, len(c.config.Mounts))
+	idmapFds := make([]int, len(c.config.Mounts))
+	for i, m := range c.config.Mounts {
+		// The -1 fd is ignored later.
+		mountFds[i] = -1
+		idmapFds[i] = -1
+		switch {
+		case m.IsBindOnly() && c.shouldSendMountSources():
```

Comment:
w.r.t. abstracting; if we inline some of the logic that's abstracted, this switch effectively becomes;

```go
		case m.IsBind() && len(m.UIDMappings) == 0 && len(m.GIDMappings) == 0 && c.shouldSendMountSources():
		case m.IsBind() && (len(m.UIDMappings) > 0 || len(m.GIDMappings) > 0) && c.shouldSendIdmapSources():
```

At least, it looks like all branches only act if `m.IsBind()`, so we could consider moving that out of the switch (and `continue` early on in the loop (_possibly_ consider moving the shared `cmd.ExtraFiles = append(cmd.ExtraFiles, messageSockPair.child)` outside of the switch as well.

If we combine some of that, and rewrite this function, it becomes something like this;

```go
func (c *Container) sendSources(cmd *exec.Cmd, messageSockPair filePair) error {
	sendMountSources := c.shouldSendMountSources()
	sendIdmapSources := c.shouldSendIdmapSources()
	if !sendMountSources && !sendIdmapSources {
		return nil
	}

	// Elements on these slices will be paired with mounts (see StartInitialization() and
	// prepareRootfs()). These slices MUST have the same size as c.config.Mounts.
	mountFds := make([]int, len(c.config.Mounts))
	idmapFds := make([]int, len(c.config.Mounts))
	for i, m := range c.config.Mounts {
		// The -1 fd is ignored later.
		mountFds[i] = -1
		idmapFds[i] = -1

		if !m.IsBind() {
			continue
		}

		// The fd passed here will not be used: nsexec.c will overwrite it with dup3(). We just need
		// to allocate a fd so that we know the number to pass in the environment variable. The fd
		// must not be closed before cmd.Start(), so we reuse messageSockPair.child because the
		// lifecycle of that fd is already taken care of.
		cmd.ExtraFiles = append(cmd.ExtraFiles, messageSockPair.child)

		if sendMountSources && !m.IsIDMapped() {
			mountFds[i] = stdioFdCount + len(cmd.ExtraFiles) - 1
		}
		if sendIdmapSources && m.IsIDMapped() {
			idmapFds[i] = stdioFdCount + len(cmd.ExtraFiles) - 1
		}
	}

	if sendMountSources {
		if err := cmdAddEnvFds(cmd, "_LIBCONTAINER_MOUNT_FDS", mountFds); err != nil {
			return err
		}
	}
	if sendIdmapSources {
		if err := cmdAddEnvFds(cmd, "_LIBCONTAINER_IDMAP_FDS", idmapFds); err != nil {
			return err
		}
	}

	return nil
}
```

Looking at that; perhaps we're trying to do too many things in this function; effectively there's two separate things happening;

- setting `_LIBCONTAINER_MOUNT_FDS` (if needed)
- setting `_LIBCONTAINER_IDMAP_FDS` (if needed)


And they both are mostly separate.

Reply:
> And they both are mostly separate.

Assuming that the order in which they get added in `cmd.ExtraFiles` doesn't matter (but let me know if they do ⚠️ ), perhaps split them to separate methods; thinking along these lines

```go
func (c *Container) sendSources(cmd *exec.Cmd, messageSockPair filePair) error {
	if err := c.sendMountSources(cmd, messageSockPair); err != nil {
		return err
	}
	if err := c.sendIdmapSources(cmd, messageSockPair); err != nil {
		return err
	}
	return nil
}

func (c *Container) sendMountSources(cmd *exec.Cmd, messageSockPair filePair) error {
	if !c.shouldSendMountSources() {
		return nil
	}

	// Elements on these slices will be paired with mounts (see StartInitialization() and
	// prepareRootfs()). These slices MUST have the same size as c.config.Mounts.
	fds := make([]int, len(c.config.Mounts))
	for i, m := range c.config.Mounts {
		// The -1 fd is ignored later.
		fds[i] = -1

		if !m.IsBind() {
			continue
		}

		// The fd passed here will not be used: nsexec.c will overwrite it with dup3(). We just need
		// to allocate a fd so that we know the number to pass in the environment variable. The fd
		// must not be closed before cmd.Start(), so we reuse messageSockPair.child because the
		// lifecycle of that fd is already taken care of.
		cmd.ExtraFiles = append(cmd.ExtraFiles, messageSockPair.child)

		if !m.IsIDMapped() {
			fds[i] = stdioFdCount + len(cmd.ExtraFiles) - 1
		}
	}

	return cmdAddEnvFds(cmd, "_LIBCONTAINER_MOUNT_FDS", fds)
}

func (c *Container) sendIdmapSources(cmd *exec.Cmd, messageSockPair filePair) error {
	if !c.shouldSendIdmapSources() {
		return nil
	}

	// Elements on these slices will be paired with mounts (see StartInitialization() and
	// prepareRootfs()). These slices MUST have the same size as c.config.Mounts.
	fds := make([]int, len(c.config.Mounts))
	for i, m := range c.config.Mounts {
		// The -1 fd is ignored later.
		fds[i] = -1

		if !m.IsBind() {
			continue
		}

		// The fd passed here will not be used: nsexec.c will overwrite it with dup3(). We just need
		// to allocate a fd so that we know the number to pass in the environment variable. The fd
		// must not be closed before cmd.Start(), so we reuse messageSockPair.child because the
		// lifecycle of that fd is already taken care of.
		cmd.ExtraFiles = append(cmd.ExtraFiles, messageSockPair.child)
		if m.IsIDMapped() {
			fds[i] = stdioFdCount + len(cmd.ExtraFiles) - 1
		}
	}

	return cmdAddEnvFds(cmd, "_LIBCONTAINER_IDMAP_FDS", fds)
}
```

Reply:
@thaJeztah I've written as different methods before and didn't like it much either, that is why I switched to this. Do you have a strong opinion? Let me know and I can change to match your taste, I didn't like much any of these, but decided for this version in the end

Reply:
I'll let others add their view as well, but my thinking here was that the separate methods may make the logic less ambiguous.

Looking at `sendIdMapSources` I can see that 

- if there's no need to send them, we return early (`! c.shouldSendIdmapSources()`)
- we're only interested in `bind` mounts
- if that bind-mount is idmapped, we're adding it to `_LIBCONTAINER_IDMAP_FDS`

Reply:
Yeap, to be honest, I was pretty much on the fence between these two options. I'll update it to have them be two different methods, then!

Reply:
@thaJeztah changed it to be different cases and different functions. But instead of just complete c&p, I've used a common function (sendFdsSources) that takes the condition as a func param.

I think this abstraction is way simpler and acceptable. But if you prefer to have it all c&p, please let me know.

Closing this, though, as it is probably better to comment in the new code (this doesn't exist). Feel free to re-open if I'm missing something